### PR TITLE
Design doc: operator based parameter server.

### PR DESCRIPTION
**Pull request by helinwang**
[Here](https://github.com/helinwang/Paddle/blob/dist_op/doc/design/ops/dist_train.md) could be easier to review

### CODE REVIEW

File: doc/design/ops/dist_train.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
+# Design Doc: Operation Graph Based Parameter Server
+
+## Abstract
+
+We propose an approach to implment the parameter server. In this
```
Comment (Superjom) [inline doc/design/ops/dist_train.md]: implement

Comment (helinwang) [reply doc/design/ops/dist_train.md]: Fixed, thanks!

File: doc/design/ops/dist_train.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+different workers.
+
+1. OPs will be added to the subgraphs, so the subgraphs can
+communicate with each other. We will need these OPs: *send*, *recv*,
+*gradient accumulator*, *string accumulator*, *loop forever*.
+
+Below is an example of converting the user defined graph to the
+sub-graphs for the trainer and the parameter server:
+
+<img src="src/local-graph.png" width="300"/>
```
Comment (jacquesqiao) [inline doc/design/ops/dist_train.md]: W is also a input of parameter update op

**Comment by putcn** [comment]
if I understand correctly, for SEND op, it only sends graph, for RECEIVE op, it only receives gradient. any given worker sees only part of the whole graph, but how would the training data travel through the whole graph? would there be some part of the graph idle until data reaches it's parent?

File: doc/design/ops/dist_train.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+After converting:
+
+<img src="src/dist-graph.png" width="500"/>
+
+1. The parameter variable W and it's optimizer subgraph are placed on the parameter server.
+1. Operators are added to the sub-graphs.
+   - *send* operator sends data and sender's address to the destination.
```
Comment (typhoonzero) [inline doc/design/ops/dist_train.md]: Question: if there are multiple parameters(or variables) to send to parameter server, are we:

1. create multiple `Send` operators for each variable or,
2. create one `Hash` operator to divide parameters equally and one `Sender` operator to do send.

Comment (typhoonzero) [reply doc/design/ops/dist_train.md]: And also, maybe add some description about the `send`, `recv` operators like:

- Send:
    - Inputs:
    - Outputs:
    - Description:

Comment (Yancey1989) [reply doc/design/ops/dist_train.md]: The same confusion with @typhoonzero , maybe we need an operator to sharded parameters.

For others, from @typhoonzero 

> create multiple Send operators for each variable

Maybe we only need one `Sender` operator. If we have too many parameters, too much `Sender` operator will cased too much connection to parameter server.

Comment (helinwang) [reply doc/design/ops/dist_train.md]: @typhoonzero @Yancey1989 sorry, the PR could be more clear:
In short, yes, "create multiple Send operators for each variable".
From the graph's perspective, the `Send` and `Recv` OP are one for each variable (but not one per replica, different (trainer) replicas share one `Send` and `Recv` for each variable).
In implementation, we could group send implementations to a single port handler.

In this design the variable placement is done by the *graph converter* before the graph is sent to worker, so it's not a runtime concept like a `Hash` operator. I think the "Hash" solution is for the simplest element-wise optimization case. If we want the parameter server to do things more than element-wise operation, we need to decide the parameter variable placement before the graph is sent to worker.

**Comment by helinwang** [comment]
@putcn Sorry if I have not explained it clearly, the `Send` OP is for sending any tensor (not the graph), the `Recv` OP is for receiving any tensor. In this way data (tensor) can travel through the graph.
Yes, the `Recv` OP will be blocked until it receives the data, so the OPs that depends on `Recv` will be idle until then.

**Comment by helinwang** [comment]
From @Superjom : graph拼起来之后应该还是一个可以用的graph。

File: doc/design/ops/dist_train.md (hunk 3 above)
Comment (typhoonzero) [reply doc/design/ops/dist_train.md]: Understood. You mean people who develop graphs, do not need to look into the implement details about how we actually send and recv variables, the graph is how the calculations flow logically. But when we build and optimize the graph, we can make the actual send operation one per trainer.

Will you add some implementation thoughts in this PR or in another one?

Comment (helinwang) [reply doc/design/ops/dist_train.md]: @typhoonzero Thanks for the reminder on the implementation thoughts! That's a good idea. I will perhaps not mention implementation detail in this PR, but create a separate issue discussing it.

**Comment by dzhwinter** [comment]
According to the current design, there is more concept need to clarify in this design doc.
1、How the scope should be implemented in a distributed environment? Every sub-graph must run with a scope to get variable value from it. How to partition the global scope? 
[here](https://github.com/PaddlePaddle/Paddle/issues/3825) is some discussion related 
2、The distributed graph need to compatible with `Block` design.

**Pull request merged**